src/protocols/bdx/BdxMessages.h

```diff
@@ -0,0 +1,240 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines structures and utility methods for working with BDX
+ *      messages.
+ */
+
+#pragma once
+
+#include <support/BitFlags.h>
+#include <support/CodeUtils.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace BDX {
+
+enum TransferControlFlags : uint8_t
+{
+    // first 4 bits reserved for version
+    kSenderDrive   = 0x10,
+    kReceiverDrive = 0x20,
+    kAsync         = 0x40,
+};
+
+enum RangeControlFlags : uint8_t
+{
+    kDefLen      = 0x01,
+    kStartOffset = 0x02,
+    kWiderange   = 0x10,
+};
+
+struct TransferInit;
```

Comment:
Instead of pre-declare, these using clauses could be after the definition below.

Reply:
I thought having them before would make it clearer to a reader that TransferInit is representing two separate messages. Should I be avoiding the forward declaration?

Reply:
I don't have an authoritative perspective on the style implications. Maybe it's more readable as-is.